app/multitenant/consul_pipe_router.go

```diff
@@ -0,0 +1,605 @@
+package multitenant
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/gorilla/mux"
+	"github.com/gorilla/websocket"
+	consul "github.com/hashicorp/consul/api"
+	"golang.org/x/net/context"
+
+	"github.com/weaveworks/scope/app"
+	"github.com/weaveworks/scope/common/mtime"
+	"github.com/weaveworks/scope/common/network"
+	"github.com/weaveworks/scope/common/xfer"
+)
+
+const (
+	gcInterval       = 30 * time.Second // we check all the pipes every 30s
+	pipeTimeout      = 1 * time.Minute  // pipes are closed when a client hasn't been connected for 1 minute
+	gcTimeout        = 10 * time.Minute // after another 10 minutes, tombstoned pipes are forgotten
+	longPollDuration = 10 * time.Second
+
+	privateAPIPort = 4444
+)
+
+var (
+	queryOptions = &consul.QueryOptions{
+		RequireConsistent: true,
+	}
+	writeOptions = &consul.WriteOptions{}
+	wsDialer     = &websocket.Dialer{}
+)
+
+// TODO deal with garbage collection
+type consulPipe struct {
+	CreatedAt, DeletedAt time.Time
+	UIEnd, ProbeEnd      string // Addrs where each end is connected
+	UIRef, ProbeRef      int    // Ref counts
+}
+
+func (c *consulPipe) toBytes() ([]byte, error) {
+	buf := bytes.Buffer{}
+	if err := json.NewEncoder(&buf).Encode(c); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func (c *consulPipe) fromBytes(bs []byte) error {
+	return json.NewDecoder(bytes.NewReader(bs)).Decode(c)
+}
```

Comment:
naming on `toBytes` and `fromBytes` seems inconsistent with json encoding elsewhere